src/crypto/crypto_hkdf.cc

```diff
@@ -103,20 +103,49 @@ bool HKDFTraits::DeriveBits(
   EVPKeyCtxPointer ctx =
       EVPKeyCtxPointer(EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, nullptr));
   if (!ctx || !EVP_PKEY_derive_init(ctx.get()) ||
-      !EVP_PKEY_CTX_hkdf_mode(ctx.get(),
-                              EVP_PKEY_HKDEF_MODE_EXTRACT_AND_EXPAND) ||
       !EVP_PKEY_CTX_set_hkdf_md(ctx.get(), params.digest) ||
-      !EVP_PKEY_CTX_set1_hkdf_salt(
-          ctx.get(), params.salt.data<unsigned char>(), params.salt.size()) ||
-      !EVP_PKEY_CTX_set1_hkdf_key(
-          ctx.get(),
-          reinterpret_cast<const unsigned char*>(params.key->GetSymmetricKey()),
-          params.key->GetSymmetricKeySize()) ||
       !EVP_PKEY_CTX_add1_hkdf_info(
           ctx.get(), params.info.data<unsigned char>(), params.info.size())) {
     return false;
   }
 
+  if (params.key->GetSymmetricKeySize() != 0) {
+    if (!EVP_PKEY_CTX_hkdf_mode(ctx.get(),
+                                EVP_PKEY_HKDEF_MODE_EXTRACT_AND_EXPAND) ||
+        !EVP_PKEY_CTX_set1_hkdf_salt(
+            ctx.get(), params.salt.data<unsigned char>(), params.salt.size()) ||
+        !EVP_PKEY_CTX_set1_hkdf_key(ctx.get(),
+                                    reinterpret_cast<const unsigned char*>(
+                                        params.key->GetSymmetricKey()),
+                                    params.key->GetSymmetricKeySize())) {
+      return false;
+    }
+  } else {
+    unsigned char temp_key[EVP_MAX_MD_SIZE];
+    unsigned int len = sizeof(temp_key);
+    if (params.salt.size()) {
+      if (HMAC(params.digest,
+               params.salt.data(),
+               params.salt.size(),
+               nullptr,
+               0,
+               temp_key,
+               &len) == nullptr) {
+        return false;
+      }
+    } else {
+      char salt[EVP_MAX_KEY_LENGTH] = {0};
+      if (HMAC(params.digest, salt, sizeof(salt), nullptr, 0, temp_key, &len) ==
```

Comment:
Isn't the salt length relevant and should it not be the output length of the digest?

Reply:
This [here](https://github.com/nodejs/node/commit/716caf7fe35aaa28d829b50a634f3586a5d4e53c#diff-940fdf5439e7036442cfcff92c16dea081d020742030a4519793afcf7f6850b1) is how my code looked initially. After a number of comments from @bnoordhuis the current state still works.

I believe I had it follow the algorithm definition at first but despite changes I myself thought should be relevant to the output, all WPT vectors still came to pass.

If you could have a look at the original code and make it not leak memory and lint-happy I would appreciate it.

Reply:
> despite changes I myself thought should be relevant to the output, all WPT vectors still came to pass.

Huh, okay, I figured it out. It happens to still work because `EVP_MAX_KEY_LENGTH` is 64 bytes, which is not greater than the block sizes of the hash functions being tested, causing OpenSSL to pad the salt to the block size and it does so with zeros, so there is no observable difference.

However, this silently relies on the fact that `EVP_MAX_KEY_LENGTH <= min(supported digest block size)`, which might break at some point.

> If you could have a look at the original code and make it not leak memory and lint-happy I would appreciate it.

You can still avoid dynamic allocation, but change `EVP_MAX_KEY_LENGTH` to `EVP_MAX_MD_SIZE`
Instead of `sizeof(salt)`, pass `EVP_MD_size(params.digest)` for the length (as you did originally). Do not assign it to an `unsigned int` in-between, unfortunately, OpenSSL uses (signed) `int` both as the return type of `EVP_MD_size` and as the argument type of `HMAC` (I believe, double-check this please).

Reply:
done, can you 👀?

Reply:
Yup, well done!